compiler/rustc_const_eval/src/interpret/terminator.rs

```diff
@@ -563,7 +562,12 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                     ty::Dynamic(..)
                 ));
                 let vtable = self.scalar_to_ptr(receiver_place.meta.unwrap_meta())?;
-                let fn_val = self.get_vtable_slot(vtable, u64::try_from(idx).unwrap())?;
+                let Some(ty::VtblEntry::Method(fn_inst)) = self.get_vtable_entries(vtable)?.get(idx).copied() else {
```

Comment:
Here again we lookup the `idx` in `get_vtable_entries`. It's a bit silly but MIR only gives us the index, not a higher-level representation of the function to call.

Reply:
Seems like a refactoring we could look into. If we expose `get_vtable_index_of_object_method` in a convenient way for codegen backends to get the runtime index into the vtable, then that looks like a fairly straight forward change. Seems borderline MCP worthy tho, maybe open an issue?

Reply:
So... what @eddyb is saying in https://github.com/rust-lang/rust/issues/99514#issuecomment-1190404713 is essentially that we can use the `DefId` from the `Virtual` to look up the right method directly and just ask the vtable for the concrete type, without generating all vtable entries and indexing into it.

Reply:
How do I proceed with that `DefId`? To call `self.resolve` I also need some substs...

Reply:
No this doesn't work, it leads to ICEs.
And indeed it makes no sense, I need to extract the type from the vtable and do something with it.

Reply:
(f26c40b0062d16dedd80f531d4b10e2de570ff33 was the bad commit that made no sense)

Reply:
Waait I should edit the subst array?? That feels like I am breaking at least 6 levels of abstraction...

Reply:
Okay after a lot of help from @eddyb this should be good now. :) And Miri detetcs [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=794a963bf0bbedcbdbe9612d5fd97b74) as UB. :D